app/models/work_package/pdf_export/gantt.rb

```diff
@@ -0,0 +1,1146 @@
+    attr_accessor :index, :rows, :columns, :lines, :text_column, :width, :height, :header_cells, :header_row_height, :group
+
+    def initialize(index, work_packages, header_cells, rows, columns, text_column, width, height, header_row_height)
+      @index = index
+      @rows = rows
+      @columns = columns
+      @work_packages = work_packages
+      @text_column = text_column
+      @width = width
+      @height = height
+      @header_cells = header_cells
+      @header_row_height = header_row_height
+      @lines = []
+      @group = nil
+      rows.each { |row| row.page = self }
+      columns.each { |column| column.page = self }
+    end
+
+    def add_line(left, right, top, bottom)
+      @lines.push({ left:, right:, top:, bottom: })
+    end
+
+    def add_lines(lines)
+      lines.each { |line| add_line(line[0], line[1], line[2], line[3]) }
+    end
+  end
+
+  class GanttDataRow
+    attr_accessor :index, :page, :work_package, :shape, :top, :left, :height, :bottom
+
+    def initialize(index, work_package, shape, left, top, height)
+      @index = index
+      @work_package = work_package
+      @shape = shape
+      @top = top
+      @left = left
+      @height = height
+      @bottom = top + height
+      @page = nil
+    end
+  end
+
+  class GanttDataColumn
+    attr_accessor :date, :left, :right, :width, :work_packages, :page
+
+    def initialize(date, left, width, work_packages)
+      @date = date
+      @left = left
+      @right = left + width
+      @width = width
+      @work_packages = work_packages
+      @page = nil
+    end
+  end
+
+  class GanttDataLineInfo
+    attr_accessor :page_group, :rows, :start_row, :start_left, :start_top, :finish_row, :finish_left, :finish_top
+
+    def initialize(page_group, rows, start_row, finish_row)
+      @page_group = page_group
+      @rows = rows
+      @start_row = start_row
+      @finish_row = finish_row
+      init_positions
+    end
+
+    def init_positions
+      @start_left = @start_row.shape.left
+      @start_top = @start_row.shape.top + (@start_row.shape.height / 2)
+      @finish_left = @finish_row.shape.right
+      @finish_top = @finish_row.shape.top + (@finish_row.shape.height / 2)
+    end
+  end
+
+  class GanttDataHeaderCell
+    attr_accessor :text, :left, :right, :top, :bottom, :height, :width
+
+    def initialize(text, left, right, top, bottom)
+      @text = text
+      @left = left
+      @right = right
+      @top = top
+      @bottom = bottom
+      @height = bottom - top
+      @width = right - left
+    end
+  end
+
+  class GanttDataTextColumn
+    attr_accessor :title, :width, :left, :right, :top, :height, :bottom, :padding_h, :padding_v
+
+    def initialize(title, left, width, top, height, padding_h, padding_v)
+      @title = title
+      @width = width
+      @left = left
+      @right = left + width
+      @padding_h = padding_h
+      @padding_v = padding_v
+      @top = top
+      @height = height
+      @bottom = top + height
+    end
+  end
```

Comment:
Immutable classes can be created in fewer lines using Data class, those that needs mutation using Struct class. Both also support keyword arguments if needed and can be combined with value omission, which will be better when there are a lot of values.

```ruby
Point = Data.define(:x, :y)
x = 1
y = 2
Point.new(x:, y:)
```

Reply:
Thanks a lot! Converted all to Data or Struct. Much much nicer.